VariantShredding.md

```diff
@@ -25,290 +25,316 @@
 
-Any values stored in a shredded `variant_value` field may have dictionary IDs referring to the metadata.
-There is one metadata value for the entire Variant record, and that is stored in the top-level `metadata` field.
-This means any `variant_value` values in the shredded representation is only the "value" portion of the [Variant Binary Encoding](VariantEncoding.md).
+The series of `tags` arrays `["comedy", "drama"], ["horror", null], ["comedy", "drama", "romance"], null` would be stored as:
 
-The metadata is kept at the top-level, instead of shredding the metadata with the shredded variant values because:
-* Simplified shredding scheme and specification. No need for additional struct-of-binary values, or custom concatenated binary scheme for `variant_value`.
-* Simplified and good performance for write shredding. No need to rebuild the metadata, or re-encode IDs for `variant_value`.
-* Simplified and good performance for Variant reconstruction. No need to re-encode IDs for `variant_value`.
+| Array                            | `value`     | `typed_value `| `typed_value...value` | `typed_value...typed_value`    |
+|----------------------------------|-------------|---------------|-----------------------|--------------------------------|
+| `["comedy", "drama"]`            | null        | non-null      | [null, null]          | [`comedy`, `drama`]            |
+| `["horror", null]`               | null        | non-null      | [null, `00`]          | [`horror`, null]               |
+| `["comedy", "drama", "romance"]` | null        | non-null      | [null, null, null]    | [`comedy`, `drama`, `romance`] |
+| null                             | `00` (null) | null          |                       |                                |
 
-# Data Skipping
+#### Objects
 
-Shredded columns are expected to store statistics in the same format as a normal Parquet column.
-In general, the engine can only skip a row group or page if all rows in the `variant_value` field are null, since it is possible for a `variant_get` expression to successfully cast a value from the `variant_value` to the target type.
-For example, if `typed_value` is of type `int64`, then the string “123” might be contained in `variant_value`, which would not be reflected in statistics, but could be retained by a filter like `where variant_get(col, “$.field”, “long”) = 123`.
-If `variant_value` is all-null, then the engine can prune pages or row groups based on `typed_value`.
-This specification is not strict about what values may be stored in `variant_value` rather than `typed_value`, so it is not safe to skip rows based on `typed_value` unless the corresponding `variant_value` column is all-null, or the engine has specific knowledge of the behavior of the writer that produced the shredded data.
+Fields of an object can be shredded using a Parquet group for `typed_value` that contains shredded fields.
 
-# Shredding Semantics
+If the value is an object, `typed_value` must be non-null.
```

Comment:
why is this required, lets say none of the fields end up getting shredded?  This is again for simplicity?

Reply:
This is a case that @cashmand pointed out above.

Say you have shredded fields `a: int`, `b: string`, and `c: double` and the projection requested projection is `object<a: int, b: string>`. The engine should be able to project the `value` and `typed_value` fields for from the groups for `a` and `b`, without projecting the object's `value`.

If the spec were to allow encoding all shredded fields as missing by setting the object's `typed_value` to null, then it wouldn't be possible to determine whether the requested projection is `null` because the value is not an object or if it is an empty object (`{ }`) because the value was an object that didn't have fields `a`, `b`, or `c`.

The solution is to use the definition level of `typed_value` to encode whether the value is an object so that it is known when you project any of the fields.

Reply:
Added a suggestion below to capture this.